waku/v2/protocol/waku_filter/protocol.nim

```diff
@@ -0,0 +1,282 @@
+  let connOpt = await wf.peerManager.dialPeer(peer, WakuFilterCodec)
+  if connOpt.isNone():
+    return err(dialFailure)
+
+  let connection = connOpt.get()
+
+  await connection.writeLP(rpc.encode().buffer)
+
+  return ok()
+
+proc sendFilterRpcToRemotePeer(wf: WakuFilter, rpc: FilterRPC, peer: RemotePeerInfo): Future[WakuFilterResult[void]] {.async, gcsafe.}=
+  let connOpt = await wf.peerManager.dialPeer(peer, WakuFilterCodec)
+  if connOpt.isNone():
+    return err(dialFailure)
+
+  let connection = connOpt.get()
+
+  await connection.writeLP(rpc.encode().buffer)
+
+  return ok()
+
+
+### Send message to subscriptors
+proc removePeerFromFailedPeersTable(wf: WakuFilter, subs: seq[Subscription]) = 
+  ## Clear the failed peer table if subscriber was able to connect
+  for sub in subs:
+    wf.failedPeers.del($sub)
+
+proc handleClientError(wf: WakuFilter, subs: seq[Subscription]) {.raises: [Defect, KeyError].} = 
+  ## If we have already failed to send message to this peer,
+  ## check for elapsed time and if it's been too long, remove the peer.
+  for sub in subs:
+    let subKey: string = $(sub)
+    
+    if not wf.failedPeers.hasKey(subKey):
+      # add the peer to the failed peers table.
+      wf.failedPeers[subKey] = Moment.now() 
+      return
+
+    let elapsedTime = Moment.now() - wf.failedPeers[subKey]
+    if elapsedTime > wf.timeout:
+      wf.failedPeers.del(subKey)
+
+      let index = wf.subscriptions.find(sub)
+      wf.subscriptions.delete(index)
+
+
+proc handleMessage*(wf: WakuFilter, pubsubTopic: string, msg: WakuMessage) {.async.} =
+  if wf.subscriptions.len <= 0:
+    return
+
+  var failedSubscriptions: seq[Subscription]
+  var connectedSubscriptions: seq[Subscription]
+
+  for sub in wf.subscriptions:
+    # TODO: Review when pubsubTopic can be empty and if it is a valid case
+    if sub.pubSubTopic != "" and sub.pubSubTopic != pubsubTopic:
+      continue
+
+    if msg.contentTopic notin sub.contentTopics:
+      continue
+
+    let rpc = FilterRPC(
+      requestId: sub.requestId,
+      push: MessagePush(messages: @[msg])
+    )
+
+    let res = await wf.sendFilterRpcToPeer(rpc, sub.peer)
+    if res.isErr():
+      waku_filter_errors.inc(labelValues = [res.error()])
+      failedSubscriptions.add(sub)
+      continue
+      
+    connectedSubscriptions.add(sub)
+
+  wf.removePeerFromFailedPeersTable(connectedSubscriptions)
+
+  wf.handleClientError(failedSubscriptions)
+
+
+### Send subscription/unsubscription
+
+proc subscribe(wf: WakuFilter, pubsubTopic: string, contentTopics: seq[ContentTopic], peer: RemotePeerInfo): Future[WakuFilterResult[string]] {.async, gcsafe.} =
+  let id = generateRequestId(wf.rng)
+  let rpc = FilterRPC(
+    requestId: id,
+    request: FilterRequest(
+      subscribe: true,
+      pubSubTopic: pubsubTopic,
+      contentFilters: contentTopics.mapIt(ContentFilter(contentTopic: it))
+    )
+  )
+
+  let res = await wf.sendFilterRpcToRemotePeer(rpc, peer)

```

Comment:
I vaguely remember that it may not be as simple as this, but why couldn't we use `sendFilterRpcToPeer()` here and just use `peer.peerId`? IIRC the `dialPeer*()` method with only a `peerId` on the peer manager is to dial existing peers - in our case, we are guaranteed that all peers we dial here already exists. Or am I missing something?

Reply:
Please, check my answer to your other comment :)